### PR TITLE
Add `molmoact2` vendor for serving MolmoAct2-DROID

### PR DESCRIPTION
**Pull request by vertix**
## Summary
Adds a `molmoact2` vendor that serves the pretrained `allenai/MolmoAct2-DROID` checkpoint in-process, mirroring the lerobot vendor (`VendorServer` subclass — no subprocess/fork):
- `codecs.py` — ordered 3-camera DROID observation (`[exterior_1, exterior_2, wrist]`, raw uint8; the model tiles to 378² itself) + raw 8-D `[q(7), grip]` state; action decode reuses `absolute_joints_action` (8-vec → `JointPosition(7)` + `target_grip`).
- `policy.py` — loads via transformers `AutoModelForImageTextToText` (`trust_remote_code`) and calls `predict_action`.
- `server.py` — single-model `VendorServer`, `cfn.cli`.
- Wired into `docker/docker-compose.yml` (`molmoact2-server`) and `workflows/nebius/serve.sh` (`molmoact2` runs on `positro/positronic` + `--extra molmoact2`).

**Dependencies:**
- New conflicting `molmoact2` extra (transformers v5; mutually exclusive with the lerobot extras). `exclude-newer` bumped 2026-05-21 → 2026-06-12 so it resolves (re-locks the other extras at the new date).
- Pins `evdev<1.9`: 1.9+ regenerates its C bindings from whatever `input-event-codes.h` is on the include path, so a newer header leaking into a cold build on the ubuntu22.04 (kernel-5.15) base fails the compile; 1.8.x reads `/usr/include` and is structurally immune (pynput needs only `evdev>=1.3`).

**Gripper convention:** MolmoAct2/DROID emits `0=open, 1=closed`, which matches positronic's real Robotiq/DH drivers, so grip passes through with no flip. Note: the MuJoCo sim's grip is inverted (`target_grip=1 → open`), so in-sim eval of this (or any DROID-convention) policy would command the gripper backwards — a pre-existing sim/real split, not handled here.

## Test plan
- End-to-end on a Nebius H100: served the model and ran a `RemotePolicy` inference returning a 15-step horizon of `JointPosition(7)` + grip with sane joint radians. (Post-review tweaks — meta keys, codec config form, the dep pin — are behavior-preserving.)
- Local codec encode/decode roundtrip via `droid.instantiate()` (3 images, 8-D state → `JointPosition(7)` + `target_grip`).
- `evdev` pin verified with real `linux/amd64` docker builds on the cuda-ubuntu22.04 base: 1.9.x fails the compile under a split header, 1.8.0 builds clean.
- `uv lock --check` clean; `molmoact2` + `lerobot` extras resolve; `evdev` → 1.8.0.
- `uv run --locked --extra dev pytest positronic/policy/tests positronic/offboard/tests` → 162 passed.

### CODE REVIEW

File: positronic/vendors/molmoact2/server.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            raise ValueError(f'Unknown model: {model_id}. Available: {self.model_id}')
+        if self._policy is None:
+            self._policy = await asyncio.to_thread(self._load_policy)
+        return self._policy, {'model_id': self.model_id}
+
+    def create_policy(self, model_handle: Any) -> Policy:
+        return model_handle
+
+    async def get_models(self) -> dict:
+        return {'models': [self.model_id]}
```
[the remark this answers is not in the document]
Comment (vertix) [reply positronic/vendors/molmoact2/server.py:69]: Fixed in 84733b1. `get_models()`/`resolve_model` now advertise a URL-safe id derived from the repo's last segment (`MolmoAct2-DROID`), while the model still loads from the full `hf_repo` (`allenai/MolmoAct2-DROID`). The advertised id is a single path segment, so a client echoing it via `RemotePolicy(model_id=...)` routes to `/api/v1/session/MolmoAct2-DROID` and matches. I scoped this to the vendor rather than switching the route to `{model_id:path}`, since that's a core-wide change to the server/client encoding.

File: positronic/vendors/molmoact2/policy.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            task=obs.get('task', ''),
+            state=np.asarray(obs['state'], dtype=np.float32),
+            norm_tag=self._norm_tag,
+            inference_action_mode='continuous',
+            enable_depth_reasoning=False,
+            num_steps=self._num_steps,
+            normalize_language=True,
+            enable_cuda_graph=False,
+        )
+        actions = out.actions[0].float().cpu().numpy()
```
[the remark this answers is not in the document]
Comment (vertix) [reply positronic/vendors/molmoact2/policy.py:34]: Checked this against the model code before changing anything: `MolmoAct2ForConditionalGeneration.predict_action` is decorated with `@torch.no_grad()` in `modeling_molmoact2.py`. So no autograd graph is built for the call (nothing to leak across websocket sessions), and `out.actions` does not require grad — `out.actions[0].float().cpu().numpy()` works without `.detach()`. The passing end-to-end run (which performs exactly that numpy conversion every call) confirms it.

Since the grad context is already handled inside `predict_action`, wrapping the call again in `torch.inference_mode()` here would be redundant, and the repo's style guidelines call out avoiding defensive guards for problems that don't actually fire — the sibling lerobot/gr00t/openpi policies likewise don't wrap, relying on their inference entrypoints. Leaving as-is; will revisit if a future checkpoint ever drops that decorator.

[the remark this answers is not in the document]
Comment (vertix) [reply positronic/vendors/molmoact2/policy.py:34]: `predict_action` is decorated `@torch.no_grad()` (see its definition in `modeling_molmoact2.py`), so the forward builds no autograd graph: the returned `out.actions` carry no grad, the `.float().cpu().numpy()` conversion can't fail on a grad-requiring tensor, and there's no graph to leak across websocket calls. An `inference_mode()` wrap would be strictly redundant here.

The autocast suggestion doesn't apply either: the model is loaded directly in `dtype=torch.bfloat16`, so the weights are already bf16 — autocast is for fp32 weights with mixed-precision compute, not for an already-bf16 module.

Documented in e415e32 so this is explicit at the call site. Leaving the thread open for you to close.

File: positronic/vendors/molmoact2/policy.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        self._processor = processor
+        self._norm_tag = norm_tag
+        self._num_steps = num_steps
+        self._meta = meta
+
+    def __call__(self, obs: dict[str, Any]) -> list[dict[str, Any]]:
+        # predict_action is decorated @torch.no_grad(), so the forward builds no autograd graph and the
+        # returned actions carry no grad — wrapping this in torch.inference_mode() or adding detach()
+        # would be redundant.
+        out = self._model.predict_action(
```
[the remark this answers is not in the document]
Comment (vertix) [reply positronic/vendors/molmoact2/policy.py:22]: Verified against the model's own code (`modeling_molmoact2.py` in the `allenai/MolmoAct2-DROID` snapshot): an external autocast context isn't needed for this bf16-loaded model.

- The model is loaded with `dtype=torch.bfloat16`, so weights and the forward run in bf16 natively — there are no fp32 weights for autocast to down-cast.
- Its attention path only consults autocast for **fp32** inputs (`if xq.dtype == torch.float32: target_dtype = torch.get_autocast_gpu_dtype() if torch.is_autocast_enabled() else self.wq.weight.dtype`, lines 1617-1621). With a bf16 module `xq` is bf16, so this branch falls back to the weight dtype regardless.
- It even explicitly **disables** autocast for numerically-sensitive ops and casts to fp32 by hand (`with torch.autocast(enabled=False, ...)`, line 148).
- `predict_action` itself is `@torch.no_grad()` (line 4502) and moves/casts its inputs internally, so the fp32 `state` we pass is handled by the model.

So wrapping the call in `torch.autocast` would be a no-op here. Documented at the call site in 80512a3. Leaving the thread open for you to close.
